### PR TITLE
HF models application integration

### LINKED ISSUE #4455: Huggingface integration - Transformers engine
### 🚩 Problem

Currently, OTX does not have native support for Hugging Face Transformers, which is a leading library for state-of-the-art NLP models and increasingly popular for vision and multimodal tasks. Integrating Hugging Face Transformers would enable OTX users to leverage a wide range of pretrained transformer models seamlessly.

---

### 💡 Proposal: Introduce `TransformersEngine` as a New Backend

We propose adding a dedicated `TransformersEngine` to OTX as a separate backend. This engine would:

- Wrap Hugging Face Transformers model instantiation, training, evaluation, and export within the OTX framework
- Allow users to utilize pretrained transformer models for vision tasks directly through OTX
- Fit into OTX’s existing CLI (stage 2)  and API design, supporting multiple interchangeable engines

---

### ✅ Goals

1. **Implement a `TransformersEngine` wrapper**  
   Integrate core Hugging Face functionality with OTX’s engine interface, enabling smooth training and inference workflows.

2. **Keep the integration modular and isolated**  
   Ensure that the `TransformersEngine` can coexist without impacting the native engine or other backends.

3. **Expose transformer-specific configurations**  
   Allow customization of model architectures, tokenizers, and training parameters through OTX configuration files and CLI.

4. **Support diverse tasks and datasets**  
 Support vision tasks, including classification, object detection, segmentation compatible with OTX datasets and metrics.

### PR DESCRIPTION
**Pull request by kprokofi**
### Summary

resolves https://github.com/open-edge-platform/geti/issues/4455

This PR introduces new models, HF weights source, configurable threshold for HF Engine and several important fixes to make new models robust

Instance Segmentation
image
image
image
image
image

Detection
image
image
image
image

Classification
image
image
image

### Checklist

- [ ] The PR title and description are clear and descriptive
- [ ] I have manually tested the changes
- [ ] All changes are covered by automated tests
- [ ] All related issues are linked to this PR (if applicable)
- [ ] Documentation has been updated (if applicable)

### CODE REVIEW

File: application/backend/app/services/model_service.py [hunk 1]
```diff
@@ -537,8 +541,11 @@ def _get_variant_binary_files(
             if metadata_file.exists():
                 paths.append(metadata_file)
             return True, tuple(paths)
-        if pt_file.exists():
-            return True, (pt_file,)
+        pytorch_files = ([pt_file] if pt_file.is_file() else []) + (
+            sorted(path for path in pt_dir.rglob("*") if path.is_file()) if pt_dir.is_dir() else []
+        )
+        if pytorch_files:
+            return True, tuple(pytorch_files)
```
Comment (leoll2) [inline application/backend/app/services/model_service.py:548]: Do the new models have multiple checkpoints?

File: application/backend/app/supported_models/manifests/classification/convnextv2_atto.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+id: image-classification-convnextv2-atto
+name: ConvNeXt V2-Atto
+license: CC BY-NC 4.0
```
Comment (leoll2) [inline application/backend/app/supported_models/manifests/classification/convnextv2_atto.yaml:3]: Convnext V2 is also available in timm, however it's currently filtered out because its code license was resolved as CC-BY-NC-4.0 in the ([csv](https://github.com/itallix/timm_license_extractor/blob/8e9036c14816cde255cee448e58d3de8ee53fd5c/timm_model_licenses.csv#L191)). The [original repo](https://github.com/facebookresearch/ConvNeXt-V2), however, is release under MIT [license](https://github.com/facebookresearch/ConvNeXt-V2#license) except for the weights which are CC-BY-NC-4.0.

In short, there are two possibilities:
1. The csv is wrong, the code license should be `mit` and we can actually expose Convnext v2 through timm.
2. The csv is right, the code license is actually CC-BY-NC-4.0 so we shouldn't expose ConvNeXt v2 at all.

What do you think? Does it make sense to support Convnext v2 - a 4y old model - as a first-class algorithm instead of through timm?

CC @itallix

File: application/backend/app/supported_models/manifests/detection/dinov3_detr_s.yaml [hunk 3]
```diff
@@ -1,5 +1,6 @@
 id: object-detection-dinov3-detr-s
 name: DINOv3-DETR-S
+license: DINOv3 License
```
Comment (leoll2) [inline application/backend/app/supported_models/manifests/detection/dinov3_detr_s.yaml]: Why this license change?

File: application/backend/app/supported_models/manifests/detection/rtdetrv2_r101.yaml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+description: "RT-DETRv2-R101 is a high-capacity real-time end-to-end detector with selective multi-scale feature extraction and improved training strategies."
+stats:
+  gigaflops: 259
+  trainable_parameters: 76.556268
+  benchmark_metrics:
+    coco_map_50_95: 54.3
+    coco_map_50: 72.8
+capabilities:
+  xai: false
+  tiling: false
```
Comment (leoll2) [inline application/backend/app/supported_models/manifests/detection/rtdetrv2_r101.yaml:16]: Why do the new algorithms not support tiling?

File: application/backend/app/supported_models/manifests/detection/rtdetrv2_r18.yaml [hunk 5]
```diff
@@ -0,0 +1,40 @@
+# Copyright (C) 2026 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+id: object-detection-rtdetrv2-r18
+name: RT-DETRv2-R18
+license: Apache 2.0
```
Comment (leoll2) [inline application/backend/app/supported_models/manifests/detection/rtdetrv2_r18.yaml]: minor: when the license is Apache 2.0, this line can be omitted.

File: application/backend/app/supported_models/manifests/classification/dinov3_vits.yaml [hunk 6]
```diff
@@ -0,0 +1,31 @@
+id: image-classification-dinov3-vits
+name: DINOv3 ViT-S/16
+license: DINOv3 License
```
Comment (leoll2) [inline application/backend/app/supported_models/manifests/classification/dinov3_vits.yaml:3]: For models that have a non-Apache2 weights license, the UI should be display a link to the actual license text. For EdgeCrafter we implemented a custom UI component, but I think we should now implement a more generic approach: extend the manifests `license` field with a `url` attribute. I created a ticket for it: https://github.com/open-edge-platform/geti/issues/7593

File: application/backend/app/supported_models/manifests/classification/convnextv2_atto.yaml (hunk 2 above)
Comment (itallix) [reply application/backend/app/supported_models/manifests/classification/convnextv2_atto.yaml:3]: It seems that the LLM didn't manage to resolve [this header](https://github.com/huggingface/pytorch-image-models/blob/9a4bc93ea7a8f1c83686bcf1695f098e81ac79fc/timm/models/convnext.py#L33-L38
) properly :

```
# ConvNeXt-V2
# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.
# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree (Attribution-NonCommercial 4.0 International (CC BY-NC 4.0))
# No code was used directly from ConvNeXt-V2, however the weights are CC BY-NC 4.0 so beware if using commercially.
```

The note `No code was used directly from ConvNeXt-V2` in the `timm` repository means that the license should actually be resolved as `Apache-2.0`. I'll need to tune the prompt and regenerate the licenses.

Comment (kprokofi) [reply application/backend/app/supported_models/manifests/classification/convnextv2_atto.yaml:3]: @leoll2 I inspected the license before adding this model, the code is indeed MIT license, but the weights are CC-BY-NC-4.0 
I see no problem to expose it in our project

The question regarding the separate model vs via timm is a good one. I added it as separate card since it is still one of the best model in HF hub, it has interesting idea behind: a fully convolutional masked autoencoder. 

However, we have DinoV3-ConvNext versions, that is distilled version from DinoV3 VIT. Problem with DinoV3 - bad license, CC-BY-NC-4.0 is also not a good one though. 

Additionally we have ConvNext-V2-atto which is much smaller than DINOV3 distilled versions.

We could then expose atto / nano versions of pure ConvNext-V2 and support only DINOV3 for base - large

File: application/backend/app/supported_models/manifests/detection/dinov3_detr_s.yaml (hunk 3 above)
Comment (kprokofi) [reply application/backend/app/supported_models/manifests/detection/dinov3_detr_s.yaml]: Let's discuss it in the OSPDT chat

File: application/backend/app/supported_models/manifests/detection/rtdetrv2_r101.yaml (hunk 4 above)
Comment (kprokofi) [reply application/backend/app/supported_models/manifests/detection/rtdetrv2_r101.yaml:16]: Well, it is a separate task, will have a ticket

File: application/backend/app/execution/common/geti_config_converter.py [hunk 7]
```diff
@@ -709,6 +716,90 @@ def _convert_ultralytics(recipe_path: Path, config: dict, hyper_parameters: dict
...
+        input_size = (
+            training_parameters.get("input_size_height"),
+            training_parameters.get("input_size_width"),
+        )
+        if all(value is not None for value in input_size):
+            config_dict["data"]["input_size"] = input_size
+            config_dict["model"]["init_args"]["data_input_params"] = {"input_size": input_size}
+
+        if (batch := training.get("batch")) is not None:
+            for subset_name in ("train_subset", "val_subset"):
```
Comment (itallix) [inline application/backend/app/execution/common/geti_config_converter.py]: Should the value for `batch_size` be applied to `test_subset` as well?

File: application/backend/app/execution/training/getitune_trainer.py [hunk 8]
```diff
@@ -468,47 +469,49 @@ def train_model(  # noqa: PLR0912, PLR0915, C901 - training orchestration is int
...
+                model_cfg["init_args"]["pretrained_weights"] = weights_path
+            elif load_from_checkpoint:
                 engine_kwargs["checkpoint"] = weights_path
                 # Disable default pretrained loading when checkpoint controls initialization.
                 model_cfg["init_args"]["pretrained"] = False
             else:
-                # Fresh Lightning training loads base weights via model init args.
+                # Fresh Lightning/HF training loads base weights via model init args.
                 model_cfg["init_args"]["pretrained"] = True
                 model_cfg["init_args"]["pretrained_weights"] = weights_path
```
Comment (itallix) [inline application/backend/app/execution/training/getitune_trainer.py]: The extra condition block duplicates the existing one. Better to merge conditions:

```suggestion
            if load_from_checkpoint and not is_huggingface:
engine_kwargs["checkpoint"] = weights_path
# Disable default pretrained loading when checkpoint controls initialization.
model_cfg["init_args"]["pretrained"] = False
else:
                # Fresh Lightning/HF training loads base weights via model init args.
                # HF base snapshots and trained revisions are both complete local
                # ``from_pretrained`` directories. Load them directly so model
                # construction never falls back to the remote architecture ID.
model_cfg["init_args"]["pretrained"] = True
model_cfg["init_args"]["pretrained_weights"] = weights_path
```

File: application/backend/app/services/model_service.py (hunk 1 above)
Comment (kprokofi) [reply application/backend/app/services/model_service.py:548]: Some new models use directory-backed Hugging Face checkpoints with multiple files, such as config.json and one or more *.safetensors files. Others remain single-file .pth checkpoints.
rglob collects all files under model/ so sharded checkpoints are stored and downloaded correctly.

File: application/backend/app/supported_models/manifests/classification/dinov3_vits.yaml (hunk 6 above)
Comment (kprokofi) [reply application/backend/app/supported_models/manifests/classification/dinov3_vits.yaml:3]: I will leave a name for now

File: application/backend/app/supported_models/manifests/detection/dinov3_detr_s.yaml (hunk 3 above)
Comment (kprokofi) [reply application/backend/app/supported_models/manifests/detection/dinov3_detr_s.yaml]: I changed it to DEIMV2 License

File: library/src/getitune/backend/huggingface/trainers/base.py [hunk 9]
```diff
@@ -194,6 +217,56 @@ def compute_loss(  # pyrefly: ignore[bad-override]
...
+            return
+        step = self.state.global_step
+        if step > self._plateau_warmup_steps:
+            self._plateau_warmup_steps = 0
+            self._set_learning_rate(self._plateau_warmup_base_lr)
+            return
+        self._set_learning_rate(plateau_warmup_lr(self._plateau_warmup_base_lr, step, self._plateau_warmup_steps))
+
+    def _set_learning_rate(self, lr: float) -> None:
+        assert self.optimizer is not None  # noqa: S101 - only called from the training loop
```
Comment (itallix) [inline library/src/getitune/backend/huggingface/trainers/base.py]: let's raise an error instead of `assert` stmt (since python byte-compiler can strip it out)

Comment (kprokofi) [reply library/src/getitune/backend/huggingface/trainers/base.py]: Sure, missed this part

File: application/backend/app/execution/common/geti_config_converter.py (hunk 7 above)
Comment (kprokofi) [reply application/backend/app/execution/common/geti_config_converter.py]: Moved to above and added test_subset

File: application/backend/app/execution/training/getitune_trainer.py (hunk 8 above)
Comment (kprokofi) [reply application/backend/app/execution/training/getitune_trainer.py]: Done

File: library/src/getitune/backend/huggingface/trainers/base.py (hunk 9 above)
Comment (kprokofi) [reply library/src/getitune/backend/huggingface/trainers/base.py]: I removed asserts from other files as well, cleaned up to not leave it for separate task

File: application/backend/app/services/base_weights_service.py [hunk 10]
```diff
@@ -194,6 +194,23 @@ def _get_huggingface_snapshot(
...
+            huggingface_hub.snapshot_download(
+                repo_id=pretrained_weights.repo_id,
+                revision=pretrained_weights.revision,
+                local_dir=local_path,
+                local_files_only=True,
+            )
+            logger.info("Using cached Hugging Face snapshot for {}", model_manifest_id)
+            return local_path
+        except LocalEntryNotFoundError:
+            logger.info("No complete local snapshot for {}; downloading.", model_manifest_id)
```
Comment (itallix) [inline application/backend/app/services/base_weights_service.py:198]: This section duplicates lines 183-194. I wonder if we can get rid of those in favor of the added?

File: library/src/getitune/backend/huggingface/models/dinov3_classification.py [hunk 11]
```diff
@@ -65,9 +65,28 @@ def __init__(
...
-        pooled = getattr(outputs, "pooler_output", None)
-        if pooled is None:
-            pooled = outputs.last_hidden_state[:, 0]
+        pooler_output = getattr(outputs, "pooler_output", None)
+        if pooler_output is not None:
+            pooled = pooler_output
+        else:
+            # Fall back through the shapes every backbone family can emit:
+            # token sequences [B, tokens, C] take the CLS token, convolutional
+            # backbones (e.g. timm ConvNeXt wrappers) give spatial [B, C, H, W]
```
Comment (itallix) [inline library/src/getitune/backend/huggingface/models/dinov3_classification.py:74]: Are timm ConvNeXt wrappers used in huggingface backend?

Comment (kprokofi) [reply library/src/getitune/backend/huggingface/models/dinov3_classification.py:74]: Yes — dinov3_convnext_* recipes load timm/convnext_*.dinov3_lvd1689m through _DinoV3Factory → AutoModel.from_config(TimmWrapperConfig) (i.e. TimmWrapperModel), whose unpooled last_hidden_state is spatial [B, C, H, W] 

The backbone we get handed is decided by the AutoConfig inside that checkpoint, not by us — and the three checkpoint distribution channels attached to the dinov3_image_classifier marker produce three different output layouts from the same AutoModel.from_config(...) call

timm/vit_*_dinov3.*
timm/convnext_*_dinov3.*
facebook/dinov3-vits16-pretrain-lvd1689m etc.

File: application/backend/app/services/base_weights_service.py (hunk 10 above)
Comment (kprokofi) [reply application/backend/app/services/base_weights_service.py:198]: Yes, removed: https://github.com/open-edge-platform/geti/pull/7581/commits/40614d1bdbd1c20896fa5110f6b4aa3f69df5dff

**Review by itallix — APPROVED** [review]
